File: src/com/cattailsw/nanidroid/shiori/NanidroidShiori.java
package com.cattailsw.nanidroid.shiori;

import java.util.HashSet;
import java.util.Hashtable;

import android.content.Context;
import java.util.Locale;
import java.io.File;
import java.io.IOException;
import java.io.FileInputStream;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.io.InputStream;
import android.util.Log;

public class NanidroidShiori extends EchoShiori {
    private static final String TAG = "NanidroidShiori";
    HashSet<String> igTable = new HashSet<String>();
    public static final String RES_NO_CONTENT = "SHIORI/3.0 204 NO CONTENT";
    private static final String RES_HEADER = "SHIORI/3.0 200 OK\r\nSender: "+ TAG +"\r\nValue: ";
    private static final String RES_END = "\r\nCharset: UTF-8\r\n";

    private static final String CONTENT_FILE_NAME = "content.txt";


    private Hashtable<String, String> evtTable = null;

    String header = null;
    private Context mCtx = null;
    private String rootpath = null;
    //    Hashtable<String, String> reqTable = null;
    
    public NanidroidShiori() {

    }

    public NanidroidShiori(Context ctx, String path) {
	mCtx = ctx;
	rootpath = path;

	// scan and get localized dir
	String userLocale = Locale.getDefault().getLanguage();
	File locDir = new File(rootpath, userLocale);
	    Log.d(TAG, "loc dir=" + locDir.getAbsolutePath());
	if ( locDir.exists() == false ) {
	    // need to find a fallback... ouch
	    Log.d(TAG, "loc dir=" + locDir.getAbsolutePath() + " not found");
	    locDir = new File(rootpath, "ja");
	}

	try {
	    readContent( new File(locDir, CONTENT_FILE_NAME) );
	}
	catch(IOException e) {

	}
	catch(Exception e){
	    e.printStackTrace();
	}
    }

    private void readContent(File contentFile) throws IOException {
	if ( contentFile.exists() ) {
	    evtTable = new Hashtable<String, String>();

	    InputStream is = new FileInputStream(contentFile);
	    BufferedReader br = new BufferedReader(new InputStreamReader(is, Charset.forName("UTF-8") ) );
	    String line = null;
	    while( true ) {
		line = br.readLine();
		if ( line == null )
			break;
		if ( line.startsWith(";") )
		    continue;
		if ( line.indexOf(',') == -1 )
		    continue; // bad line

		int idx = line.indexOf(',');
		String key = line.substring(0,idx);
		String val = line.substring(idx + 1);
		evtTable.put(key, val);
	    }
	}
    }

    public void terminate() {

    }

    public String getModuleName() {
	return "NanidroidShiori";
    }
    
    
    @Override
    protected String genResponse() {
	if ( reqTable == null  || mCtx == null )
	    return super.genResponse();
	
	String reqId = reqTable.get("id");
	if ( reqId.equalsIgnoreCase("OnGhostChanging")) {
	    return  handleGhostChanging();//RES_HEADER + reqTable.get("id") + "\\e";
	}
	else if ( reqId.equalsIgnoreCase("OnGhostChanged")) {
	    return handleGhostChanged();
	}
	else if ( reqId.equalsIgnoreCase("OnClose") ) {
	    return handleOnClose();
	}
	else if ( reqId.equalsIgnoreCase("OnBoot")) {
	    return handleOnBoot();
	}
	else if ( reqId.equalsIgnoreCase("OnInstallFailure") ) {
	    return handleInstallFail();
	}
		

	if ( evtTable.get(reqId) != null )
	    return RES_HEADER + evtTable.get(reqId) + RES_END;

	return RES_NO_CONTENT;
	//	return super.genResponse();
    }

    private String handleGhostChanging() {
	if ( evtTable.get("OnGhostChanging") != null ) {
	    String s = evtTable.get("OnGhostChanging");    
	    return RES_HEADER + String.format(s, reqTable.get("reference0")) + RES_END;
	}
	
	return RES_NO_CONTENT;
    }

    private String handleGhostChanged() {
	if ( evtTable.get("OnGhostChanged") != null ) {
	    String s = evtTable.get("OnGhostChanged");
	    return RES_HEADER + String.format(s, reqTable.get("reference0")) + RES_END;
	}

	return RES_NO_CONTENT;// RES_HEADER + reqTable.get("reference0") + RES_END;
    }

    private String handleOnClose() {
	if ( evtTable.get("OnClose") != null ) {
	    String s = evtTable.get("OnClose");
	    return RES_HEADER + s + RES_END;
	}

	return RES_HEADER + "OnClose" + RES_END;
    }

    private String handleInstallFail() {
	if ( evtTable.get("OnInstallFailure") != null ) {

	    return RES_HEADER + evtTable.get("OnInstallFailure") + RES_END;
	}
	
	return RES_NO_CONTENT;
    }

    private String handleOnBoot() {
	if ( evtTable.get("OnBoot") != null ) {
	    return RES_HEADER + evtTable.get("OnBoot") + RES_END;
	}
	return RES_NO_CONTENT;
    }

}

File: src/com/cattailsw/nanidroid/shiori/NanidroidShiori.kt
package com.cattailsw.nanidroid.shiori

import android.content.Context
import android.util.Log
import java.io.BufferedReader
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStreamReader
import java.nio.charset.Charset
import java.util.Hashtable
import java.util.Locale

open class NanidroidShiori() : EchoShiori() {
    private var evtTable: Hashtable<String, String>? = null
    private var mCtx: Context? = null
    private var rootpath: String? = null

    constructor(ctx: Context?, path: String) : this() {
        mCtx = ctx
        rootpath = path

        val userLocale = Locale.getDefault().language
        var locDir = File(rootpath, userLocale)
        Log.d(TAG, "loc dir=${locDir.absolutePath}")
        if (!locDir.exists()) {
            Log.d(TAG, "loc dir=${locDir.absolutePath} not found")
            locDir = File(rootpath, "ja")
        }

        try {
            readContent(File(locDir, CONTENT_FILE_NAME))
        } catch (_: IOException) {
            // Legacy behavior intentionally ignores unreadable content files.
        } catch (exception: Exception) {
            exception.printStackTrace()
        }
    }

    @Throws(IOException::class)
    private fun readContent(contentFile: File) {
        if (contentFile.exists()) {
            evtTable = Hashtable()

            val input = FileInputStream(contentFile)
            val reader = BufferedReader(InputStreamReader(input, Charset.forName("UTF-8")))
            while (true) {
                val line = reader.readLine() ?: break
                if (line.startsWith(";")) continue
                val index = line.indexOf(',')
                if (index == -1) continue

                evtTable!![line.substring(0, index)] = line.substring(index + 1)
            }
        }
    }

    override fun terminate() = Unit

    override fun getModuleName(): String = "NanidroidShiori"

    override fun genResponse(): String {
        val values = reqTable ?: return super.genResponse()
        if (mCtx == null) return super.genResponse()

        // The Java implementation throws when a request is missing ID. Keep that
        // observable parser behavior instead of silently treating it as 204.
        val requestId = values["id"]!!
        return when {
            requestId.equals("OnGhostChanging", ignoreCase = true) -> handleGhostChanging()
            requestId.equals("OnGhostChanged", ignoreCase = true) -> handleGhostChanged()
            requestId.equals("OnClose", ignoreCase = true) -> handleOnClose()
            requestId.equals("OnBoot", ignoreCase = true) ||
                requestId.equals("OnFirstBoot", ignoreCase = true) -> handleOnBoot()
            requestId.equals("OnInstallFailure", ignoreCase = true) -> handleInstallFail()
            evtTable!![requestId] != null -> RES_HEADER + evtTable!![requestId] + RES_END
            else -> RES_NO_CONTENT
        }
    }

    private fun handleGhostChanging(): String {
        val script = evtTable!!["OnGhostChanging"] ?: return RES_NO_CONTENT
        return RES_HEADER + String.format(script, reqTable!!["reference0"]) + RES_END
    }

    private fun handleGhostChanged(): String {
        val script = evtTable!!["OnGhostChanged"] ?: return RES_NO_CONTENT
        return RES_HEADER + String.format(script, reqTable!!["reference0"]) + RES_END
    }

    private fun handleOnClose(): String {
        val script = evtTable!!["OnClose"] ?: return RES_HEADER + "OnClose" + RES_END
        return RES_HEADER + script + RES_END
    }

    private fun handleInstallFail(): String {
        val script = evtTable!!["OnInstallFailure"] ?: return RES_NO_CONTENT
        return RES_HEADER + script + RES_END
    }

    private fun handleOnBoot(): String {
        val script = evtTable!!["OnBoot"] ?: return RES_NO_CONTENT
        return RES_HEADER + script + RES_END
    }

    companion object {
        private const val TAG = "NanidroidShiori"

        @JvmField
        val RES_NO_CONTENT = "SHIORI/3.0 204 NO CONTENT"

        private const val RES_HEADER = "SHIORI/3.0 200 OK\r\nSender: $TAG\r\nValue: "
        private const val RES_END = "\r\nCharset: UTF-8\r\n"
        private const val CONTENT_FILE_NAME = "content.txt"
    }
}
